### DISCUSSION TITLE
how to implement the new version?

### DISCUSSION THREAD

**Discussion by t4tohd** [Novage/p2p-media-loader#401] [General]
Hello, I was using the old version and when I replaced the two updated libraries it no longer loads anything, I use clappr, the demo version in Novage does not load, I have used these libraries and they do not work for me:
https://cdn.jsdelivr.net/npm/p2p-media-loader-core@1.0.3/dist/p2p-media-loader-core.es.min.js
https://cdn.jsdelivr.net/npm/p2p-media-loader-hlsjs@1.0.3/dist/p2p-media-loader-hlsjs.es.min.js
https://cdn.jsdelivr.net/npm/@clappr/player@~0/dist/clappr.min.js
https://cdn.jsdelivr.net/gh/clappr/clappr-level-selector-plugin@~0/dist/level-selector.min.js

Could you update the guides to implement the new version  .

<script src="https://cdn.jsdelivr.net/npm/p2p-media-loader-core@0.6.2/build/p2p-media-loader-core.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/p2p-media-loader-hlsjs@0.6.2/build/p2p-media-loader-hlsjs.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/clappr@latest"></script>
        <style>
            #video {
                width: 720px;
                margin-left: auto;
                margin-right: auto;
                display: block;
            }
        </style>
         <div id="video"></div>
        <script>
            if (p2pml.hlsjs.Engine.isSupported()) {
                var engine = new p2pml.hlsjs.Engine();
                var player = new Clappr.Player({
                    parentId: "#video",
                    source: "https://cph-p2p-msl.akamaized.net/hls/live/2000341/test/master.m3u8",
                    playback: {
                        hlsjsConfig: {
                            liveSyncDurationCount: 7, // To have at least 7 segments in queue
                            loader: engine.createLoaderClass(),
                        },
                    },
                });
                p2pml.hlsjs.initClapprPlayer(player);
                player.setVolume(1); // 1%
                player.play();
            } else {
                document.write("Not supported :(");
            }
        </script>

**Comment by DimaDemchenko** [comment]
Hello, @t4tohd


Thank you for submitting the issue. I'll review our demo and fix the problem with the Clappr player.

Regarding your implementation:
We have rewritten the p2p-media-loader, and the previous version is not compatible with new one. Please refer to our [p2p-media-loader Documentation](https://novage.github.io/p2p-media-loader/docs/latest/) to learn how to use and configure the new p2p loader.

Please take a look and feel free to provide us with feedback about the documentation. We would like to know if it was informative, clear to read, and if there is anything we should add for clarity.


Here is a very simple working example with Clappr player:

```
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Document</title>

    <script
      type="text/javascript"
      src="https://cdn.jsdelivr.net/npm/@clappr/player@~0/dist/clappr.min.js"
    ></script>
    <script
      type="text/javascript"
      src="https://cdn.jsdelivr.net/gh/clappr/clappr-level-selector-plugin@~0/dist/level-selector.min.js"
    ></script>

    <script type="importmap">
      {
        "imports": {
          "p2p-media-loader-core": "https://cdn.jsdelivr.net/npm/p2p-media-loader-core@^1/dist/p2p-media-loader-core.es.min.js",
          "p2p-media-loader-hlsjs": "https://cdn.jsdelivr.net/npm/p2p-media-loader-hlsjs@^1/dist/p2p-media-loader-hlsjs.es.min.js"
        }
      }
    </script>
  </head>
  <body>
    <style>
      #video {
        width: 720px;
        height: 480px;
        margin-left: auto;
        margin-right: auto;
        display: block;
      }
    </style>
    <div id="video"></div>

    <script type="module">
      import { HlsJsP2PEngine } from "p2p-media-loader-hlsjs";

      const streamUrl =
        "https://cph-p2p-msl.akamaized.net/hls/live/2000341/test/master.m3u8";

      const engine = new HlsJsP2PEngine();

      const clapprPlayer = new Clappr.Player({
        source: streamUrl,
        plugins: [LevelSelector],
        height: "100%",
        width: "100%",
        parentId: `#video`,
        playback: {
          hlsjsConfig: {
            ...engine.getConfigForHlsJs(),
          },
        },
      });

      engine.bindHls(() => clapprPlayer.core.getCurrentPlayback()?._hls);

      clapprPlayer.setVolume(1);
    </script>
  </body>
</html>
```

**Comment by ElonJTrump** [comment]
Any chance for JWplayer ?

**Comment by mrlika** [comment]
Our investigation shows that there is no good external hls.js provider for JWPlayer. Its built-in hls.js provider isn't configurable\patchable at all. Unfortunately, there is no good news...

**Comment by ghost** [comment]
> Our investigation shows that there is no good external hls.js provider for JWPlayer. Its built-in hls.js provider isn't configurable\patchable at all. Unfortunately, there is no good news...

hello,
can you help you me implement the new version with PlayerJS ?

```
<!DOCTYPE html>
<html>
    <head>
        <meta charset="utf-8">
        <meta name="viewport" content="width=device-width, initial-scale=1">
        <meta name="robots" content="none">
        <title>Video</title>
        <script src="/assets/js/jquery-3.6.0.min.js"></script>
        <script src="playerjs/assets/js/playerjs_default_19.10.5.js"></script>
	<script src="https://cdn.jsdelivr.net/npm/p2p-media-loader-core@1.0.3/dist/p2p-media-loader-core.es.min.js"></script>
	<script src="https://cdn.jsdelivr.net/npm/p2p-media-loader-hlsjs@1.0.3/dist/p2p-media-loader-hlsjs.es.min.js"></script>
	<style>html,body{margin:0;padding:0;overflow:hidden;width:100%;height:100%}#video_player{width:100%;height:100%}</style>
    </head>
    <body>
        <div id="video_player"></div>
<script type="text/javascript">
        var Get_P2P = 0;
        if (Get_P2P && p2pml.hlsjs.Engine.isSupported()) {
            var engine = new p2pml.hlsjs.Engine({
                segments: {
                    swarmId: "xxxxxxxxx",
                },
                loader: {
                    trackerAnnounce: ["custom_tracker"],
                    // rtcConfig: {iceServers: [{urls: ["stun:stun.l.google.com:19302", "stun:global.stun.twilio.com:3478"]}]},
                    cachedSegmentExpiration: 86400000,
                    cachedSegmentsCount: 1000,
                    // requiredSegmentsPriority: 6,
    			}
            });
        }
    	var ID = "xxxxxxxxx";
    	var player = new Playerjs({
    	    id: "video_player",
    	    file: "link.m3u8",
    	    poster: "link.jpg",
    	    subtitle_start: 1,
    	    subtitle: "",
    	    download: "",
    	    thumbnails: "link.vtt",
    	    autoplay: 0,
    	    lang: "en",
            hlsconfig: {
                loader: engine.createLoaderClass(),
            },
    	    default_audio: "English",
    	});
    
    if (Get_P2P && p2pml.hlsjs.Engine.isSupported()) {
        p2pml.hlsjs.initHlsJsPlayer(player.api('hls'));
        var downloaded_total = 0;
        var downloaded = 0;
        engine.on('piece_bytes_downloaded', function (method, segment, bytes, peerId) {
            if(peerId){
                downloaded += bytes;
            }
            downloaded_total += bytes;
        });
        // setInterval(function(){ console.log('Download from P2P: ' + ((downloaded*100)/downloaded_total) + '%'); }, 5000);
    }
    </script>
    </body>
</html>
```

**Comment by ghost** [comment]
@mrlika @DimaDemchenko 
hello, i found debug in console.log but i can't found wss tracker load in Network tab. what should i do?

![image](https://github.com/user-attachments/assets/aa34ab58-3614-4d89-b5ff-c77bb8c5703b)

```
	<script type="module">
	import { HlsJsP2PEngine } from "p2p-media-loader-hlsjs";
	var Get_P2P = 1;
	if (Get_P2P) {
		var hlsP2PEngine = new HlsJsP2PEngine({
			core: {
				swarmId: "8XZVWVH2uyqN",
				announceTrackers: ["wss://custom1", "wss://custom2"],
			}
		});
	}
	var player = new Playerjs({
	....
		hlsconfig: {
			...hlsP2PEngine.getConfigForHlsJs(),
		},
	});
	.....
	hlsP2PEngine.bindHls(() => player.api('hls'));
	</script>
```

**Comment by DimaDemchenko** [comment]
@jazz1611 I'm currently looking into your issue with implementing P2P in PlayerJS and will give you an update soon.

**Comment by ghost** [comment]
> @jazz1611 I'm currently looking into your issue with implementing P2P in PlayerJS and will give you an update soon.

thank you for support. this is PlayerJS library have included HLS.js library too. you can rely on the HTML sample above to make PlayerJS work.

`<script src="https://cdn.plrjs.com/player/ttd5tyt396tfa/nnszdu40osmd.js"></script>`

**Comment by DimaDemchenko** [comment]
@jazz1611 So, here is the update on your situation. I used a different method to implement P2P with PlayerJS. 

```
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Playerjs</title>
    <script src="playerjs.js"></script>

    <!-- Import map for P2P Media Loader modules -->
    <script type="importmap">
      {
        "imports": {
          "p2p-media-loader-core": "https://cdn.jsdelivr.net/npm/p2p-media-loader-core@^1/dist/p2p-media-loader-core.es.min.js",
          "p2p-media-loader-hlsjs": "https://cdn.jsdelivr.net/npm/p2p-media-loader-hlsjs@^1/dist/p2p-media-loader-hlsjs.es.min.js"
        }
      }
    </script>
  </head>
  <body>
    <div id="player"></div>

    <script type="module">
      import { HlsJsP2PEngine } from "p2p-media-loader-hlsjs";

      window.Hls = HlsJsP2PEngine.injectMixin(window.Hls);

      function initPlayer() {
        const player = new Playerjs({
          id: "player",
          hlsconfig: {
            // Hls.js config parameters go here
            p2p: {
              core: {
                swarmId: "Optional custom swarm ID for stream",
                // Other P2P engine config parameters go here
              },
              onHlsJsCreated: (hls) => {
                // Subscribe to P2P engine and Hls.js events here
                hls.p2pEngine.addEventListener("onPeerConnect", (details) => {
                  console.log(`Connected to peer ${details.peerId})`);
                });
              },
            },
          },
        });
      }

      document.addEventListener("DOMContentLoaded", initPlayer);
    </script>
  </body>
</html>

```

![image](https://github.com/user-attachments/assets/2ae9ea81-faa3-4b23-86b5-e7a5fa4cc627)
![image](https://github.com/user-attachments/assets/3fe0a88c-efd7-487e-89e6-03eca979a5cd)

**Comment by DimaDemchenko** [comment]
@jazz1611 I tried to run your code and it seems like it works. However, I ran it with my PlayerJS (with the HLS module enabled) because I couldn't load your script for some reason.
![image](https://github.com/user-attachments/assets/f9e8d5ea-d260-4567-a965-a7b7810ca53b)

**Comment by ghost** [comment]
@DimaDemchenko 
hello, i just make some test and it still work when use only PlayerJS library have included HLS.js library (no need external url load from hls.js)

there is what is changed.  i will make test with VideoJS soon. thank you so much.

test online: https://jsfiddle.net/nmgsr79o/

```
<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta name="robots" content="none">
    <title>Video</title>
    <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.7.1/jquery.min.js"></script>
    <script src="https://cdn.plrjs.com/player/ttd5tyt396tfa/nnszdu40osmd.js"></script>
    <script type="importmap"> {
        "imports": {
            "p2p-media-loader-core": "https://cdn.jsdelivr.net/npm/p2p-media-loader-core@latest/dist/p2p-media-loader-core.es.js",
            "p2p-media-loader-hlsjs": "https://cdn.jsdelivr.net/npm/p2p-media-loader-hlsjs@latest/dist/p2p-media-loader-hlsjs.es.js"
        }
    }
    </script>
    <style>
      html,
      body {
        margin: 0;
        padding: 0;
        overflow: hidden;
        width: 100%;
        height: 100%
      }

      #video_player {
        width: 100%;
        height: 100%
      }
    </style>
  </head>
  <body>
    <div id="video_player"></div>
    <script type="module">
      import {
        HlsJsP2PEngine
      } from "p2p-media-loader-hlsjs";
      window.Hls = HlsJsP2PEngine.injectMixin(window.Hls);
      var downloaded = 0;
      var downloaded_total = 0;
      var ID = "xxxxxxxxx";
      var player = new Playerjs({
        id: "video_player",
        file: "https://flipfit-cdn.akamaized.net/flip_hls/6656423247ffe600199e8363-15125d/video_h1.m3u8",
        poster: "link.jpg",
        subtitle_start: 1,
        subtitle: "",
        download: "",
        thumbnails: "link.vtt",
        autoplay: 0,
        lang: "en",
        hlsconfig: {
          p2p: {
            core: {
              swarmId: "xxxxxxxxx",
            },
            onHlsJsCreated: (hls) => {
              // Subscribe to P2P engine and Hls.js events here
              hls.p2pEngine.addEventListener("onPeerConnect", (details) => {
                console.log(`Connected to peer ${details.peerId})`);
              });
              hls.p2pEngine.addEventListener('onChunkDownloaded', (bytesLength, downloadSource, peerId) => {
                console.log(`Downloaded ${bytesLength} bytes from ${downloadSource} ${peerId ? 'from peer ' + peerId : 'from server'}`);
                if (peerId) {
                  downloaded += bytesLength;
                }
                downloaded_total += bytesLength;
              });
            },
          },
        },
        default_audio: "English",
      });
    </script>
  </body>
</html>
```

**Comment by ghost** [comment]
> @jazz1611 I tried to run your code and it seems like it works. However, I ran it with my PlayerJS (with the HLS module enabled) because I couldn't load your script for some reason. ![image](https://private-user-images.githubusercontent.com/91938357/349403889-f9e8d5ea-d260-4567-a965-a7b7810ca53b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjEyMDMwMTAsIm5iZiI6MTcyMTIwMjcxMCwicGF0aCI6Ii85MTkzODM1Ny8zNDk0MDM4ODktZjllOGQ1ZWEtZDI2MC00NTY3LWE5NjUtYTdiNzgxMGNhNTNiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzE3VDA3NTE1MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWE5YmU4NTk1OGMyZjg4ZDU1MzliOTdjMzkxOWY0ODlmZWNlMzEwNTEwMzUxNzYyMmVkZmI5YjQyMDljNmQ2MjkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.S8zZwHzHE2V8W47F9jKACWCW0qlW8Xw0AIGv3JlYxIs)

I have updated the exact URL so you can test online (https://jsfiddle.net/nmgsr79o/). I don't enable HLS version in PlayerJS (by default PlayerJS always adds it in its own library) I have set up P2P-Media-Loader's default tracker. Please try and let me know the results
![image](https://github.com/user-attachments/assets/43e12fec-232d-4299-bdb4-3ff16facadf3)

Another problem is that if I open 2 tabs at the same time, in JSFiddle's Console and also in the browser, it returns results continuously causing the browser to crash... I'm pretty sure P2P is working but is there a bug in the console?

![image](https://github.com/user-attachments/assets/326914b8-9d16-4595-9264-31267681d3ec)

**Comment by DimaDemchenko** [comment]
@jazz1611 
I'm not responsible for JSFiddle lags. 
I tested provided code locally and it works

![image](https://github.com/user-attachments/assets/c1d8092b-99bf-486d-a48f-0c1a86724815)

**Comment by ghost** [comment]
@DimaDemchenko 
What I mean is that it returns results in the console continuously, right? Because if only one tab is opened, it returns the correct loading results from HTTP. If there is a peer, it returns results continuously.

**Comment by DimaDemchenko** [comment]
@jazz1611 I probably understand what you mean. If there are no peers in the network, you will only download high-demand segments (15 seconds by default). If there are peers in the network, they will try to download segments within the p2pDownloadTimeWindow and httpDownloadTimeWindow. That's why you see a lot of download / upload logs. Check the [stream config documentation](
https://novage.github.io/p2p-media-loader/docs/v1.0/types/p2p_media_loader_core.StreamConfig.html ), we described those parameters here

**Comment by ghost** [comment]
@DimaDemchenko i found the script hls.js plugin with latest version for VideoJS and i do config similar from PlayerJS to VideoJS but it not work. can help you me to check it?

Source: https://www.nuevodevel.com/nuevo/showcase/hlsjs
Script: https://jsfiddle.net/a3zdv1tn/

P/s: also any chance i can run p2p library in html page with inside <script type="text/javascript">....</script>. which `type="module"` some my code not work. and im not expert with Javascript code

**Reply by mrlika** [reply]
Nuevo plugin uses Hls.js. You can likely override default Hls.js with P2P enabled Hls.js, similar to another players. Just run this code before creating Videojs with Nuevo plugin:

```
import { HlsJsP2PEngine } from "p2p-media-loader-hlsjs";
window.Hls = HlsJsP2PEngine.injectMixin(window.Hls);
```

**Comment by ElonJTrump** [comment]
I didn't try version 1.0.3 yet I try previous release and found problem with iPad OS with Apple CPU M series. Can you all share me your experience about this device with version 1.0.3.

**Comment by DimaDemchenko** [comment]
@stevenking888, we are experiencing problems with Safari. I'm currently working on version 1.0.4, which might fix these issues.

**Comment by ghost** [comment]
@DimaDemchenko 
I have successfully incorporated Hls.js library into VideoJS for P2P usage. This is the configuration that I tested and succeeded... but there is a problem that the plugins show the quality menu won't show, I'm still looking for a way to fix it.

Script: https://raw.githubusercontent.com/jazz1611/Hls.js-P2P-Media-Loader/main/videojs-hlsjs-plugin_v1.0.js

```
<video id="video_player" class="video-js vjs-default-skin vjs-big-play-centered vjs-fill" crossOrigin="anonymous"></video>
<script type="module">
import { HlsJsP2PEngine } from "p2p-media-loader-hlsjs";
window.Hlsjs = HlsJsP2PEngine.injectMixin(window.Hls);
var config = {
	width: "100%",
	height: "100%",
	controls: true,
	poster: "link.jpg",
	preload: "auto",
	controlBar: {
		skipButtons: {
			forward: 5,
			backward: 5
		},
		volumePanel: {
			inline: false
		}
	},
	enableDocumentPictureInPicture: true,
	qualityLevels: true,
	// playbackRates: [0.25, 0.5, 0.75, 1, 1.25, 1.5, 1.75, 2],
	sources: "linkstream.m3u8",
	html5: {
		hlsjsConfig: {
			p2p: {
				core: {
					swarmId: "xxxxxxxx",
					// Other P2P engine config parameters go here
				},
				onHlsJsCreated: (hls) => {
					// Subscribe to P2P engine and Hls.js events here
					hls.p2pEngine.addEventListener("onPeerConnect", (details) => {
						console.log(`Connected to peer ${details.peerId})`);
					});
				},
			},
		},
	},
};
var player = videojs("video_player", config);
</script>
```

**Reply by t4tohd** [reply]
Do you recommend any configuration that can be done in nginx/conf/nginx.conf to extend the stability of streaming and the p2p download range?

**Comment by mrlika** [comment]
P2P download ratio doesn't depend on Nginx configuration

**Reply by ghost** [reply]
@mrlika Is there a way to run P2P in HTML5, without using `type="module"` in the script tag. if use like this then P2P not working.

Example:

```
<script type="module">
import { HlsJsP2PEngine } from "p2p-media-loader-hlsjs";
window.Hlsjs = HlsJsP2PEngine.injectMixin(window.Hls);
</script>
<script>
my custom script
</script>
```

**Reply by mrlika** [reply]
Unfortunately, no. We decided to use modern browser features supported by 92% of browsers. You can detect browser features and avoid loading of P2P functionality in browsers without modules/importmap support.

Another option is to use the library as NPM module and bundle it into an application.

**Reply by mrlika** [reply]
You can detect importmap support like:

```javascript
function detectImportmapSupport() {
  return HTMLScriptElement.supports && HTMLScriptElement.supports('importmap');
}

function loadModules() {
  if (detectImportmapSupport()) {
    // Use importmap
    const importMap = document.createElement('script');
    importMap.type = 'importmap';
    importMap.textContent = JSON.stringify({
      imports: {
        // Define your module mappings here
        'module-name': '/path/to/module.js'
      }
    });
    document.head.appendChild(importMap);

    // Use the modules
    const module = await import('module-name');
  } else {
    // Fallback for browsers without importmap support
    // ...
  }
}

loadModules();
```

**Reply by ghost** [reply]
@DimaDemchenko, @mrlika thank you for support. also i have also successfully integrated P2P into JW Player and it works but now there is a confusing error... my video is HLS VOD

The first segments download with difficulty and when I watch it it freezes at the 9th or 10th second. Sometimes it freezes/hangs at the 8th minute or so. unstable... and one more bug is that it downloads all video segments when a peer is detected even though the playback buffer is only for the next 60 seconds

PlayerJS, VideoJS and JW Player all have the above problem. How should I set it up correctly? 

```
swarmId: "custom_ID",
announceTrackers: ["wss://tracker1", "wss://tracker2"],
// simultaneousHttpDownloads: 2,
// p2pNotReceivingBytesTimeoutMs: 2000,
// httpNotReceivingBytesTimeoutMs: 3000,
highDemandTimeWindow: 60,
cachedSegmentExpiration: 21600000,
```

**Reply by DimaDemchenko** [reply]
@jazz1611 Regarding 

> and one more bug is that it downloads all video segments when a peer is detected even though the playback buffer is only for the next 60 seconds

It's not a bug; it's expected behavior. The parameter `httpDownloadTimeWindow` is configured to download 3000 seconds (default value) of video segments in advance if there is at least one peer, and this one peer doesn't have those segments. Similarly, we have `p2pDownloadTimeWindow`, which is configured to download 6000 seconds (default value) of video segments in advance over the P2P network.

To sum up, if there are peers in the network, we attempt to request 6000 seconds of video segments in advance from those peers. If those peers don't have the requested segments, we then attempt to download 3000 seconds of video segments through HTTP in advance to be able to share the downloaded segments with peers. If there are no peers in the network, you will only download segments that are located in the `highDemandTimeWindow`.

Regarding

> The first segments download with difficulty and when I watch it it freezes at the 9th or 10th second. Sometimes it freezes/hangs at the 8th minute or so. unstable...

Please provide us with more information.

- Does this issue only occur with your VOD HLS stream?
- Is your bandwidth sufficient to download the segments quickly?
- What is your segment length and size? 
- provide the code of your integrations, so I can test

To gather more details on what's happening with your segment requests, please type the following command in your browser console, restart the page, and monitor the output:
`localStorage.debug = 'p2pml-core:request-main'`

**Reply by ghost** [reply]
> Does this issue only occur with your VOD HLS stream?

Yes

> Is your bandwidth sufficient to download the segments quickly?

Yes

> What is your segment length and size?

Segment length is 10 second and avg size 1-2MB/file

> To gather more details on what's happening with your segment requests, please type the following command in your browser console, restart the page, and monitor the output:
> localStorage.debug = 'p2pml-core:request-main'

```
p2pml-core:request-main http 0 succeed +704ms
Connected to peer -PM0100-sMNSexd6s2FE)
p2pml-core:request-main p2p 2 started +0ms
p2pml-core:request-main p2p 2 succeed +135ms
Connected to peer -PM0100-5JA37xAvTRjQ)
p2pml-core:request-main p2p 4 started +0ms
p2pml-core:request-main http 5 started +0ms
p2pml-core:request-main p2p 4 succeed +141ms
p2pml-core:request-main p2p 3 started +0ms
p2pml-core:request-main http 8 succeed +421ms
p2pml-core:request-main p2p 10 started +0ms
p2pml-core:request-main p2p 10 succeed +127ms
..........
p2pml-core:request-main p2p 18 started +0ms
p2pml-core:request-main p2p 18 succeed +46ms
p2pml-core:request-main http 21 started +0ms
p2pml-core:request-main http 22 started +0ms
p2pml-core:request-main http 21 succeed +282ms
p2pml-core:request-main http 22 succeed +426ms
p2pml-core:request-main http 24 started +0ms
p2pml-core:request-main http 28 started +0ms
p2pml-core:request-main p2p 23 started +0ms
p2pml-core:request-main p2p 23 succeed +38ms
p2pml-core:request-main http 32 succeed +371ms
p2pml-core:request-main http 34 succeed +433ms
..........
p2pml-core:request-main http 40 started +0ms
p2pml-core:request-main http 39 succeed +280ms
p2pml-core:request-main p2p 36 started +0ms
p2pml-core:request-main p2p 36 succeed +124ms
p2pml-core:request-main p2p 38 started +0ms
p2pml-core:request-main http 40 succeed +711ms
p2pml-core:request-main p2p 38 succeed +127ms
```

This is configure code for JW Player

```
<script type="importmap">
        {
            "imports": {
                "p2p-media-loader-core": "//cdn.jsdelivr.net/npm/p2p-media-loader-core@latest/dist/p2p-media-loader-core.es.min.js",
                "p2p-media-loader-hlsjs": "//cdn.jsdelivr.net/npm/p2p-media-loader-hlsjs@latest/dist/p2p-media-loader-hlsjs.es.min.js"
            }
        }
</script>
<script src="//ssl.p.jwpcdn.com/player/v/8.34.3/jwplayer.js"></script>
<script src="https://raw.githubusercontent.com/jazz1611/Hls.js-P2P-Media-Loader/main/jwplayer_provider-hlsjs.js"></script>
<div id="video_player"></div>
<script type="module">
        import { HlsJsP2PEngine } from "p2p-media-loader-hlsjs";
        var engine = new HlsJsP2PEngine({
              core: {
                swarmId: "xxxxxxx",
                // Other P2P engine config parameters go here
              },
        });
    var player = jwplayer("video_player");
    var config = {
        width: "100%",
        height: "100%",
        hlshtml: "true",
        preload: "auto",
        primary: "html5",
        sources: [{"file":"link.m3u8","label":"Auto","type":"hls"}],
        hlsjsConfig: {
            ...engine.getConfigForHlsJs(),
        }
    };
    player.setup(config);
    engine.bindHls(() => player.hls);
    engine.addEventListener("onPeerConnect", (details) => {
      console.log(`Connected to peer ${details.peerId})`);
    });
</script>
```
@DimaDemchenko Is there any way I can communicate directly with you?

**Reply by ghost** [reply]
https://cdn.jsdelivr.net/node_modules/.pnpm/debug@4.3.5/node_modules/debug/src/common.js

```
p2pml-core:request-main http 0 started +0ms common.js:113:9
p2pml-core:request-main http 1 started +0ms common.js:113:9
p2pml-core:request-main http 1 succeed +3s common.js:113:9
p2pml-core:request-main http 2 started +0ms common.js:113:9
p2pml-core:request-main http 0 succeed +3s common.js:113:9
p2pml-core:request-main http 2 succeed +1s common.js:113:9
p2pml-core:request-main http 1 started +0ms common.js:113:9
p2pml-core:request-main http 2 started +0ms common.js:113:9
p2pml-core:request-main http 1 succeed +1s common.js:113:9
p2pml-core:request-main http 2 succeed +2s common.js:113:9
p2pml-core:request-main http 3 started +0ms common.js:113:9
p2pml-core:request-main http 3 succeed +9s common.js:113:9
p2pml-core:request-main http 4 started +0ms common.js:113:9
p2pml-core:request-main http 4 succeed +7s common.js:113:9
p2pml-core:request-main http 5 started +0ms common.js:113:9
HLS.js error: networkError - fatal: false - fragLoadError 2 videojs-hlsjs-plugin_v1.js:14414:20
HLS.js error: networkError - fatal: false - fragLoadError videojs-hlsjs-plugin_v1.js:14414:20
HLS.js error: networkError - fatal: false - fragLoadError videojs-hlsjs-plugin_v1.js:14414:20
HLS.js error: networkError - fatal: false - fragLoadError videojs-hlsjs-plugin_v1.js:14414:20
HLS.js error: networkError - fatal: false - fragLoadError videojs-hlsjs-plugin_v1.js:14414:20
HLS.js error: networkError - fatal: true - fragLoadError videojs-hlsjs-plugin_v1.js:14414:20
bubbling network error up to VIDEOJS videojs-hlsjs-plugin_v1.js:14416:18
VIDEOJS: ERROR: (CODE:2 MEDIA_ERR_NETWORK) HLS.js error: networkError - fatal: true - fragLoadError 
Object { code: 2, message: "HLS.js error: networkError - fatal: true - fragLoadError" }
video.min.js:12:1415
p2pml-core:request-main http 6 started +0ms common.js:113:9
p2pml-core:request-main http 6 succeed +5s common.js:113:9
p2pml-core:request-main http 7 started +0ms common.js:113:9
p2pml-core:request-main http 7 failed bytes-receiving-timeout +3s common.js:113:9
p2pml-core:request-main http 7 started +0ms common.js:113:9
p2pml-core:request-main http 7 failed bytes-receiving-timeout +3s common.js:113:9
p2pml-core:request-main http 7 started +0ms common.js:113:9
p2pml-core:request-main http 7 failed bytes-receiving-timeout +3s common.js:113:9
p2pml-core:request-main http 5 succeed +39s
```

**Reply by ghost** [reply]
Another problem is that P2P does not work with manifest content as shown below. It only downloads segments according to range (206 status) but cannot watch the video.

```
#EXTM3U
#EXT-X-VERSION:4
#EXT-X-TARGETDURATION:5
#EXT-X-MEDIA-SEQUENCE:0
#EXT-X-PLAYLIST-TYPE:VOD
#EXTINF:5.046700,
#EXT-X-BYTERANGE:545388@0
https://vod-url/19978.ts?token=42AP1vO6qjcNblKi80nHIA&expires=1721761953
#EXTINF:5.005000,
#EXT-X-BYTERANGE:326556@545388
https://vod-url/19978.ts?token=42AP1vO6qjcNblKi80nHIA&expires=1721761953
#EXTINF:5.005000,
..............................
..............................
..............................
#EXTINF:5.005000,
#EXT-X-BYTERANGE:535800@106465716
https://vod-url/19978.ts?token=42AP1vO6qjcNblKi80nHIA&expires=1721761953
#EXTINF:0.333600,
#EXT-X-BYTERANGE:46812@107001516
https://vod-url/19978.ts?token=42AP1vO6qjcNblKi80nHIA&expires=1721761953
#EXT-X-ENDLIST
```
![image](https://github.com/user-attachments/assets/597c1e5d-1054-4d77-8de5-4c0a3b21208b)

**Reply by DimaDemchenko** [reply]
@jazz1611 
Please ensure that your streams can be played with the latest version of HLS. You can test this using the HLS.js demo at the following link: [Hls.Js Demo](https://hlsjs.video-dev.org/demo/)

**Reply by ghost** [reply]
> @jazz1611 Please ensure that your streams can be played with the latest version of HLS. You can test this using the HLS.js demo at the following link: [Hls.Js Demo](https://hlsjs.video-dev.org/demo/)

I tried my stream through the HLS.js demo as you mentioned and it still watch normal. but when applied to P2P, it cannot be watched
